### ISSUE TITLE
Core dump with test files and multi-backend

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/osm2pgsql#334]
with 

`test_output_multi_tags.json`

``` json
[
  {
    "name": "test_points_1",
    "type": "point",
    "tagtransform": "tests/test_output_multi_tags.lua",
    "tagtransform-node-function": "test_nodes_1",
    "tagtransform-way-function": "drop_all",
    /* No relations in the file */
    "tagtransform-relation-function": "drop_all",
    "tagtransform-relation-member-function": "drop_all",
    "tags": [
      {"name": "foo", "type": "text"},
      {"name": "bar", "type": "text"},
      {"name": "baz", "type": "text"} /* left empty by the transform */
    ]
  },
  {
    /* This second table will process the same data, but resulting in different tags */
    "name": "test_points_2",
    "type": "point",
    "tagtransform": "tests/test_output_multi_tags.lua",
    "tagtransform-node-function": "test_nodes_2",
    "tagtransform-way-function": "drop_all",
    /* No relations in the file */
    "tagtransform-relation-function": "drop_all",
    "tagtransform-relation-member-function": "drop_all",
    "tags": [
      {"name": "foo", "type": "text"},
      {"name": "bar", "type": "text"}, /* left empty by the transform */
      {"name": "baz", "type": "text"}
    ]
  }
]
```

`test_output_multi_tags.lua`

``` lua
function drop_all (...)
  return 1, {}
end

function test_nodes_1 (kv, num_tags)
  if kv["foo"] then
    tags = {}
    tags["bar"] = kv["foo"]
    return 0, tags
  else
    return 1, {}
  end
end

function test_nodes_2 (kv, num_tags)
  if kv["foo"] then
    tags = {}
    tags["baz"] = kv["foo"]
    return 0, tags
  else
    return 1, {}
  end
end
```

`test_output_multi_tags.osm`

``` xml
<osm version="0.6">
<!--
 43
 21
-->
 <node id="1" lat="49.00" lon="-123.00">
 <tag k="foo" v="n1"/>
 </node>
 <node id="2" lat="49.00" lon="-123.01"/>
 <node id="3" lat="49.01" lon="-123.00"/>
 <node id="4" lat="49.01" lon="-123.01"/>
 <way id="1">
  <nd ref="1"/>
  <nd ref="2"/>
 <tag k="foo" v="w1"/>
 </way>
 <way id="2">
  <nd ref="1"/>
  <nd ref="3"/>
  <nd ref="4"/>
  <nd ref="1"/>
 <tag k="foo" v="w2"/>
 </way>
</osm>
```

when I ran the command `./osm2pgsql -d osm2pgsqltest -O multi -S tests/test_output_multi_tags.json tests/test_output_multi_tags.osm` I got the output

```
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Sorting data and creating indexes for test_points_1
Sorting data and creating indexes for test_points_2
node cache: stored: 4(100.00%), storage efficiency: 0.39% (dense blocks: 1, sparse nodes: 0), hit rate: nan%
Copying test_points_2 to cluster by geometry finished
Creating geometry index on  test_points_2
Copying test_points_1 to cluster by geometry finished
Creating geometry index on  test_points_1
Creating indexes on  test_points_2 finished
Creating indexes on  test_points_1 finished
Abort trap (core dumped)
```

A backtrace puts it at

```
  0x0000000802eee6ca in thr_kill () from /lib/libc.so.7
[New Thread 804007400 (LWP 101038/osm2pgsql)]
[New Thread 804007000 (LWP 101037/osm2pgsql)]
[New Thread 804006400 (LWP 100843/osm2pgsql)]
(gdb) bt
#0  0x0000000802eee6ca in thr_kill () from /lib/libc.so.7
#1  0x0000000802fc3149 in abort () from /lib/libc.so.7
#2  0x000000080284442d in __cxa_rethrow () from /lib/libcxxrt.so.1
#3  0x00000000004810e6 in pgsql_exec_simple (sql_conn=<value optimized out>, expect=<value optimized out>, sql=0x7fffff7fb6e9 "ANALYZE test_points_1")
    at pgsql.cpp:44
#4  0x0000000000480eef in pgsql_exec_simple (sql_conn=0x6, expect=PGRES_EMPTY_QUERY, sql=<value optimized out>) at pgsql.cpp:32
#5  0x0000000000489fde in table_t::stop (this=0x8040a4280) at table.cpp:256
#6  0x000000000046438b in output_multi_t::stop (this=0x804038120) at output-multi.cpp:147
#7  0x00000008023634a5 in boost::thread::start_thread_noexcept () from /usr/local/lib/libboost_thread.so.1.55.0
#8  0x0000000802c654f5 in pthread_create () from /lib/libthr.so.3
#9  0x0000000000000000 in ?? ()
```

The C++ is

``` c++
boost::shared_ptr<PGresult> pgsql_exec_simple(PGconn *sql_conn, const ExecStatusType expect, const char *sql)
{
    PGresult* res;
#ifdef DEBUG_PGSQL
    fprintf( stderr, "Executing: %s\n", sql );
#endif
    res = PQexec(sql_conn, sql);
    if (PQresultStatus(res) != expect) {
        PQclear(res);
        throw std::runtime_error((boost::format("%1% failed: %2%\n") % sql % PQerrorMessage(sql_conn)).str());
    }
    return boost::shared_ptr<PGresult>(res, &PQclear);
}
```

The line it's got a core dump on is the throw. Probably related to #298 in cause of the PG error.

**Comment by mmd-osm** [comment]
Cannot reproduce this on 0.96.0-RC1 anymore.

<details>
<summary>Full log (click to open)</summary>

```
~/osm2pgsql$ build/osm2pgsql -d osm -O multi -S tests/test_output_multi_tags.json tests/test_output_multi_tags.osm
osm2pgsql version 0.96.0-RC1 (64 bit id space)

Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using projection SRS 3857 (Spherical Mercator)
Setting up table: test_points_1
Setting up table: test_points_2
Setting up table: test_line_1
Setting up table: test_line_2
Setting up table: test_polygon_1
Setting up table: test_polygon_2
Allocating memory for dense node cache
Allocating dense node cache in one big chunk
Allocating memory for sparse node cache
Sharing dense sparse
Node-cache: cache=800MB, maxblocks=12800*65536, allocation method=3

Reading in file: tests/test_output_multi_tags.osm
Using XML parser.
  parse time: 0s
Node stats: total(4), max(4) in 0s
Way stats: total(2), max(2) in 0s
Relation stats: total(0), max(0) in 0s
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua
Using lua based tag processing pipeline with script tests/test_output_multi_tags.lua

Going over pending ways...
	4 ways are pending

Using 8 helper-processes
Finished processing 4 ways in 0 s

Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2
Committing transaction for test_points_1
Committing transaction for test_points_2
Committing transaction for test_line_1
Committing transaction for test_line_2
Committing transaction for test_polygon_1
Committing transaction for test_polygon_2

Going over pending relations...
	0 relations are pending

Using 8 helper-processes
Finished processing 0 relations in 0 s

Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Committing transaction for test_points_1
WARNING:  there is no transaction in progress
Committing transaction for test_points_2
WARNING:  there is no transaction in progress
Committing transaction for test_line_1
WARNING:  there is no transaction in progress
Committing transaction for test_line_2
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_1
WARNING:  there is no transaction in progress
Committing transaction for test_polygon_2
WARNING:  there is no transaction in progress
Sorting data and creating indexes for test_points_1
Sorting data and creating indexes for test_points_2
Sorting data and creating indexes for test_line_1
Sorting data and creating indexes for test_line_2
node cache: stored: 4(100.00%), storage efficiency: 0.05% (dense blocks: 1, sparse nodes: 0), hit rate: 100.00%
Sorting data and creating indexes for test_polygon_1
Sorting data and creating indexes for test_polygon_2
Copying test_line_1 to cluster by geometry finished
Creating geometry index on test_line_1
Creating indexes on test_line_1 finished
All indexes on test_line_1 created in 0s
Completed test_line_1
Copying test_points_1 to cluster by geometry finished
Creating geometry index on test_points_1
Creating indexes on test_points_1 finished
All indexes on test_points_1 created in 0s
Completed test_points_1
Copying test_line_2 to cluster by geometry finished
Creating geometry index on test_line_2
Copying test_polygon_1 to cluster by geometry finished
Creating geometry index on test_polygon_1
Copying test_polygon_2 to cluster by geometry finished
Creating geometry index on test_polygon_2
Copying test_points_2 to cluster by geometry finished
Creating geometry index on test_points_2
Creating indexes on test_line_2 finished
Creating indexes on test_polygon_1 finished
Creating indexes on test_polygon_2 finished
Creating indexes on test_points_2 finished
All indexes on test_line_2 created in 0s
Completed test_line_2
All indexes on test_polygon_1 created in 0s
Completed test_polygon_1
All indexes on test_polygon_2 created in 0s
Completed test_polygon_2
All indexes on test_points_2 created in 0s
Completed test_points_2

Osm2pgsql took 2s overall
```

</details>

**Comment by lonvia** [comment]
Got most likely fixed with #890.
